src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+            if (collectionType.TypeKind == TypeKind.Interface && knownCollectionInterfaces.Contains(collectionType.OriginalDefinition))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
+                if (indexer != null && semanticFact.IsAssignableTo(indexer.ReturnType, foreachType, model.Compilation))
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+            }
+
+            // check regular cases (implicitly implemented)
+            string explicitInterface = null;
+            foreach (var current in collectionType.AllInterfaces)
+            {
+                if (!knownCollectionInterfaces.Contains(current.OriginalDefinition))
+                {
+                    continue;
+                }
+
+                // see how the type implements the interface
+                var countSymbol = GetInterfaceMembers(current, get_Count);
+                var indexerSymbol = GetInterfaceMembers(current, get_Item);
+                if (countSymbol == null || indexerSymbol == null)
+                {
+                    continue;
+                }
+
+                var countImpl = collectionType.FindImplementationForInterfaceMember(countSymbol) as IMethodSymbol;
+                var indexerImpl = collectionType.FindImplementationForInterfaceMember(indexerSymbol) as IMethodSymbol;
+                if (countImpl == null || indexerImpl == null)
+                {
+                    continue;
+                }
+
+                if (!semanticFact.IsAssignableTo(indexerImpl.ReturnType, foreachType, model.Compilation))
+                {
+                    continue;
+                }
+
+                // implicitly implemented!
+                if (countImpl.ExplicitInterfaceImplementations.IsEmpty &&
+                    indexerImpl.ExplicitInterfaceImplementations.IsEmpty)
+                {
+                    explicitCast = null;
+                    countName = Count;
+                    return;
+                }
+
+                if (explicitInterface == null)
+                {
+                    explicitInterface = current.ToDisplayString();
+                }
+            }
+
+            // okay, we don't have implicitly implemented one, but we do have explicitly implemented one
+            if (explicitInterface != null)
+            {
+                explicitCast = explicitInterface;
+                countName = Count;
+            }
+        }
+
+        private static bool IsNullOrErrorType(ITypeSymbol type)
+        {
+            return type == null || type is IErrorTypeSymbol;
+        }
+
+        private static IMethodSymbol GetInterfaceMembers(ITypeSymbol interfaceType, string memberName)
+        {
+            var members = interfaceType.GetMembers(memberName);
+            if (!members.IsEmpty)
+            {
+                return (IMethodSymbol)members[0];
```

Comment:
how do you know it's a method?

Reply:
because I use it only for method.

Reply:
How do you know?  What if the lookup finds a property with that name?

i.e. you do this:

```c#
+            if (collectionType.TypeKind == TypeKind.Interface && knownCollectionInterfaces.Contains(collectionType.OriginalDefinition))
+            {
+                var indexer = GetInterfaceMembers(collectionType, get_Item);
```

Your collection type could be a type that has a property called "get_Item" on it, right?

Reply:
well, it does it only for known interfaces but sure, one can define its own IList interface with same fully qualified name if get_Item property.